xarray/backends/rasterio_.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+
+try:
+    import rasterio
+except ImportError:
+    rasterio = False
+
+from .. import Variable, DataArray
+from ..core.utils import FrozenOrderedDict, Frozen, NDArrayMixin
+from ..core import indexing
+from ..core.pycompat import OrderedDict, suppress
+
+from .common import AbstractDataStore
+
+__rio_varname__ = 'raster'
+
+
+class RasterioArrayWrapper(NDArrayMixin):
+    def __init__(self, ds):
+        self._ds = ds
+        self.array = ds.read()
+
+    @property
+    def dtype(self):
+        return np.dtype(self._ds.dtypes[0])
+
+    def __getitem__(self, key):
+        if key == () and self.ndim == 0:
+            return self.array.get_value()
+        return self.array[key]
```

Comment:
By the time indexers get here, they should have already passed through [canonicalize_indexer](https://github.com/pydata/xarray/blob/a247d9314d6a8df34df887106e234bbffebe1e48/xarray/core/indexing.py#L41), which means you only need to handle a `key` that is a tuple of the appropriate number of dimensions (i.e., 3) composed of integers, slices and integer ndarrays.

Based on the [docstring for read](https://github.com/mapbox/rasterio/blob/a5376390829be52490a1fe6da52395262d0f6167/rasterio/_io.pyx#L234-L287), we want something like:

```python
def __getitem__(self, key):
    band = range(self.shape[0])[key[0]]
    window = []
    for win in key[1:]:
        if instance(win, slice):
            window.append((win.start, win.stop))
        elif isinstance(win, int):
            window.append((win, win + 1))
        else:  # integer ndarray
            window.append((win.min(), win.max()))
    raw_data = self._ds.read(band, window=window)
    # now, fix up raw_data to conform to numpy indexing conventions
    # - drop axes for integer band/windows
    # - stride if windows are slices with win.step != 1
    # - subset if window is an integer ndarray
```

Honestly, this logic should probably live in rasterio if possible. I'm a little surprised that they have never implemented a `__getitem__` method.